test/run_portability_test.py

```diff
@@ -0,0 +1,199 @@
+"""Runs a single portability test from a test directory."""
+
+import argparse
+import datetime
+import glob
+import json
+import os
+import sys
+import time
+
+import requests
+
+# This is what we need to find in a test directory. I guess we'll glob the PR
+# test WDL itself since it's named after the pipeline.
+TEST_DIR_LAYOUT = {
+    "inputs": "test_inputs.json",
+    "dependencies": "dependencies.json",
+    "test": "*PR.wdl"
+}
+
+# These are the environment variables we're expecting to be able to submit a
+# portability test.
+ENV_VARIABLES = {
+    "portability_service_url": os.environ.get("PORTABILITY_SERVICE_URL"),
+    "portability_service_headers": os.environ.get("PORTABILITY_SERVICE_HEADERS")
+}
+
+class TestFailure(Exception):
+    """Error for a test that fails its portability test."""
+    pass
+
+class TestDefinitionError(Exception):
+    """Error for malformed test directory."""
+    pass
+
+class TestEnvironmentError(Exception):
+    """Error for malformed test environment."""
+    pass
+
+def gather_test_inputs(test_dir):
+    """Walk through the test directory, finding files and dependencies needed
+    to submit a portability test.
+
+    Raise a TestDefinitionError if anything goes wrong.
+    """
+
+    errors = []
+
+    # Test WDL
+    wdl_glob = os.path.join(test_dir, TEST_DIR_LAYOUT["test"])
+    wdl_paths = glob.glob(wdl_glob)
+    if not wdl_paths:
+        errors.append("Test definition WDL not found.")
+    if len(wdl_paths) > 1:
+        errors.append("Multiple candidate test WDLs found: {}".format(wdl_paths))
+
+    try:
+        test_wdl_string = open(wdl_paths[0]).read()
+    except IOError:
+        test_wdl_string = None
+        errors.append("Test WDL {} could not be read".format(wdl_paths[0]))
+
+    # Test inputs
+    try:
+        inputs_json_path = os.path.join(test_dir, TEST_DIR_LAYOUT["inputs"])
+        inputs_json_string = open(inputs_json_path).read()
+    except IOError:
+        inputs_json_string = None
+        errors.append("Inputs JSON {} could not be read".format(inputs_json_path))
+
+    # Dependencies
+
+    # First try to load the dependency JSON itself
+    try:
+        dependencies_json_path = os.path.join(test_dir, TEST_DIR_LAYOUT["dependencies"])
+        dependencies_json_string = open(dependencies_json_path).read()
+    except IOError:
+        dependencies_json_string = None
+        errors.append("Dependencies JSON {} could not be read".format(dependencies_json_path))
+
+    # Now iterate over the dependencies and load the files.
+    loaded_dependencies = []
+    dependencies_dict = json.loads(dependencies_json_string)
+
+    for key, value in dependencies_dict.items():
+        try:
+            loaded_dependencies.append({
+                "name": key,
+                "code": open(value).read()
+            })
+        except IOError:
+            errors.append("Count not read dependency {}".format(value))
+
+    if errors:
```

Comment:
I like this -- waiting to throw errors until you need to could produce a nice productivity boost for the user.